tasmota/tasmota_xsns_sensor/xsns_100_ina3221.ino

```diff
@@ -359,30 +444,72 @@ void Ina3221Show(bool json)
+          dtostrfd(Ina3221Data[device].chan[chan].energy_ah, Settings->flag2.energy_resolution, temp);
+          strncat(energy_ah, temp, sizeof(energy_ah));
+          #endif
+          #ifdef INA3221_CALC_ENERGY_WH
+          dtostrfd(Ina3221Data[device].chan[chan].energy_wh, Settings->flag2.energy_resolution, temp);
+          strncat(energy_wh, temp, sizeof(energy_wh));
+          #endif
         } //if enabled
         else {
           strncat(voltage, "null", sizeof(voltage));
-          strncat(current, "null", sizeof(voltage));
-          strncat(power, "null", sizeof(voltage));
+          strncat(current, "null", sizeof(current));
+          strncat(power, "null", sizeof(power));
+          #ifdef INA3221_CALC_CHARGE_AH
+          strncat(energy_ah, "null", sizeof(energy_ah));
+          #endif
+          #ifdef INA3221_CALC_ENERGY_WH
+          strncat(energy_wh, "null", sizeof(energy_wh));
+          #endif
         }
         if (0xFE & enabled_chan) {
           strncat(voltage, ",", sizeof(voltage));
-          strncat(current, ",", sizeof(voltage));
-          strncat(power, ",", sizeof(voltage));
+          strncat(current, ",", sizeof(current));
+          strncat(power, ",", sizeof(power));
+          #ifdef INA3221_CALC_CHARGE_AH
+          strncat(energy_ah, ",", sizeof(energy_ah));
+          #endif
+          #ifdef INA3221_CALC_ENERGY_WH
+          strncat(energy_wh, ",", sizeof(energy_wh));
+          #endif
         }
       } // for channel
-      ResponseAppend_P(PSTR(",\"%s\":{\"Id\":\"0x%02x\",\"" D_JSON_VOLTAGE "\":[%s],\"" D_JSON_CURRENT "\":[%s],\"" D_JSON_POWERUSAGE "\":[%s]}"),
+      #if defined(INA3221_CALC_CHARGE_AH) || defined(INA3221_CALC_ENERGY_WH)
+        #if defined(INA3221_CALC_CHARGE_AH) && defined(INA3221_CALC_ENERGY_WH)
+        ResponseAppend_P(PSTR(",\"%s\":{\"Id\":\"0x%02x\",\"" D_JSON_VOLTAGE "\":[%s],\"" D_JSON_CURRENT "\":[%s],\"" D_JSON_POWERUSAGE "\":[%s],\"" D_JSON_CHARGE "\":[%s],\"" D_JSON_ENERGY "\":[%s]}"),
+                      name, Ina3221Data[device].i2caddr, voltage, current, power, energy_ah, energy_wh);
+        #else
+          #ifdef INA3221_CALC_CHARGE_AH
+          ResponseAppend_P(PSTR(",\"%s\":{\"Id\":\"0x%02x\",\"" D_JSON_VOLTAGE "\":[%s],\"" D_JSON_CURRENT "\":[%s],\"" D_JSON_POWERUSAGE "\":[%s],\"" D_JSON_CHARGE "\":[%s]}"),
+                      name, Ina3221Data[device].i2caddr, voltage, current, power, energy_ah, energy_wh);
+          #endif
+          #ifdef INA3221_CALC_ENERGY_WH
+          ResponseAppend_P(PSTR(",\"%s\":{\"Id\":\"0x%02x\",\"" D_JSON_VOLTAGE "\":[%s],\"" D_JSON_CURRENT "\":[%s],\"" D_JSON_POWERUSAGE "\":[%s],\"" D_JSON_ENERGY "\":[%s]}"),
+                      name, Ina3221Data[device].i2caddr, voltage, current, power, energy_ah, energy_wh);
+          #endif
+        #endif
+      #else
+        ResponseAppend_P(PSTR(",\"%s\":{\"Id\":\"0x%02x\",\"" D_JSON_VOLTAGE "\":[%s],\"" D_JSON_CURRENT "\":[%s],\"" D_JSON_POWERUSAGE "\":[%s]}"),
```

Comment:
too many repetition makes the code hard to maintain
Please refactor

Reply:
ok I see .... but Ihave no idea how to do it to have one or the other or both itemms in the strig or display ?? :-((

Reply:
I've not tried to compile but that should do the job
```
      ResponseAppend_P(PSTR(",\"%s\":{\"Id\":\"0x%02x\",\"" D_JSON_VOLTAGE "\":[%s],\"" D_JSON_CURRENT "\":[%s],\"" D_JSON_POWERUSAGE "\":[%s]"
      #if INA3221_CALC_CHARGE_AH
        ",\"" D_JSON_CHARGE "\":[%s]"
      #endif
      #ifdef INA3221_CALC_ENERGY_WH
        ",\"" D_JSON_ENERGY "\":[%s]"
      #endif
      "}", name, Ina3221Data[device].i2caddr, voltage, current, power
      #if INA3221_CALC_CHARGE_AH
        , energy_ah
      #endif
      #ifdef INA3221_CALC_ENERGY_WH
        , energy_wh
      #endif
      );
```

Reply:
I edited the code above as there was a closing parenhtesis missing after the end of the string  in line
```
    "}"), name, Ina3221Data[device].i2caddr, voltage, current, power
```